snapcraft/tests/test_sources.py

```diff
@@ -169,6 +170,45 @@ def test_extract_and_keep_debfile(self):
             self.assertEqual('Test fake compressed file', deb_file.read())
 
 
+class TestRpm(tests.TestCase):
+
+    def test_pull_rpm_file_must_extract(self):
+        rpm_file_name = 'test.rpm'
+        dest_dir = 'src'
+        os.makedirs(dest_dir)
+
+        test_file_path = os.path.join(self.path, 'test.txt')
+        open(test_file_path, 'w').close()
+        rpm_file_path = os.path.join(self.path, rpm_file_name)
+        os.chdir(self.path)
+        with libarchive.file_writer(rpm_file_path, 'cpio', 'gzip') as rpm:
+            rpm.add_files('test.txt')
+
+        rpm_source = sources.Rpm(rpm_file_path, dest_dir)
+        rpm_source.pull()
+
+        self.assertEqual(os.listdir(dest_dir), ['test.txt'])
+
+    def test_extract_and_keep_rpmfile(self):
+        rpm_file_name = 'test.rpm'
+        dest_dir = 'src'
+        os.makedirs(dest_dir)
+
+        test_file_path = os.path.join(self.path, 'test.txt')
+        open(test_file_path, 'w').close()
+        rpm_file_path = os.path.join(self.path, rpm_file_name)
+        os.chdir(self.path)
+        with libarchive.file_writer(rpm_file_path, 'cpio', 'gzip') as rpm:
+            rpm.add_files('test.txt')
+
+        rpm_source = sources.Rpm(rpm_file_path, dest_dir)
+        rpm_source.provision(dst=dest_dir, keep_rpm=True)
+        rpm_source.pull()
```

Comment:
You shouldn't call pull after provision. That will result on provision being called twice.